e2e/internal/e2e/home_linux.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package e2e
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"testing"
+
+	"github.com/sylabs/singularity/internal/pkg/buildcfg"
+	"github.com/sylabs/singularity/internal/pkg/util/user"
+)
+
+// rpmMacrosContent contains required content to
+// place in $HOME/.rpmmacros file for yum bootstrap
+// build
+var rpmMacrosContent = `
+%_var /var
+%_dbpath %{_var}/lib/rpm
+`
+
+// SetupHomeDirectories creates temporary home directories for
+// privileged and unprivileged users and bind mount those directories
+// on top of real ones. It's possible because e2e tests are executed
+// in a dedicated mount namespace.
+func SetupHomeDirectories(t *testing.T) {
+	var unprivUser, privUser *user.User
+
+	sessionDir := buildcfg.SESSIONDIR
+	unprivUser = CurrentUser(t)
+
+	Privileged(func(t *testing.T) {
+		// there is no cleanup here because everything done (tmpfs, mounts)
+		// in our dedicated mount namespace will be automatically discarded
+		// by the kernel once all test processes exit
+
+		privUser = CurrentUser(t)
+
+		// create the temporary filesystem
+		if err := syscall.Mount("tmpfs", sessionDir, "tmpfs", 0, "mode=0777"); err != nil {
+			t.Fatalf("failed to mount temporary filesystem")
+		}
+
+		// want the already resolved current working directory
+		cwd, err := os.Readlink("/proc/self/cwd")
+		if err != nil {
+			t.Fatalf("could not readlink /proc/self/cwd: %s", err)
+		}
+		unprivRealHome, err := filepath.EvalSymlinks(unprivUser.Dir)
+		if err != nil {
+			t.Fatalf("could not resolve %s: %s", unprivUser.Dir, err)
+		}
+		privRealHome, err := filepath.EvalSymlinks(privUser.Dir)
+		if err != nil {
+			t.Fatalf("could not resolve %s: %s", privUser.Dir, err)
+		}
+
+		// prepare user temporary homes
+		unprivHome := filepath.Join(sessionDir, unprivUser.Name)
+		privHome := filepath.Join(sessionDir, privUser.Name)
+
+		oldUmask := syscall.Umask(0)
+		defer syscall.Umask(oldUmask)
+
+		if err := os.Mkdir(unprivHome, 0700); err != nil {
+			t.Fatalf("failed to create temporary home %s: %s", unprivHome, err)
+		}
+		if err := os.Chown(unprivHome, int(unprivUser.UID), int(unprivUser.GID)); err != nil {
+			t.Fatalf("failed to set temporary home %s owner: %s", unprivHome, err)
+		}
+		if err := os.Mkdir(privHome, 0700); err != nil {
+			t.Fatalf("failed to create temporary home %s: %s", privHome, err)
+		}
+
+		// re-create current working directory if tests are run from the user
+		// home directory and bind it
+		if strings.HasPrefix(cwd, unprivRealHome) {
```

Comment:
> correct?
so it's extracting (part of) the current working directory relative to the home directory of the current (?) user and reparenting on top of the temporary home?

Right

> And why is this necessary only if the current working directory is within the current user's home? (why is there a condition on line 83)

As we use go module, the source could be everywhere, if the source base is in `/tmp/singularity` we don't need to bind CWD on top of the user home directory, but if source code is indeed somewhere in user directory we need to bind it on top of the temporary one in order to execute tests correctly.

> Also, what's special about the directory "3" above? Why it is kept out of this whole thing?

It would have been easier to have something like `buildcfg.SOURCEDIR` in order to determine the source code location, but as tests are executed relative to the directory where `_test.go` is located, so in `$SOURCEDIR/e2e`, `filepath.Dir(cwd)` is equivalent to source directory

Reply:
> As we use go module, the source could be everywhere, if the source base is in `/tmp/singularity` we don't need to bind CWD on top of the user home directory, but if source code is indeed somewhere in user directory we need to bind it on top of the temporary one in order to execute tests correctly.

Ah, got it. That makes sense.

> > Also, what's special about the directory "3" above? Why it is kept out of this whole thing?
> 
> It would have been easier to have something like `buildcfg.SOURCEDIR` in order to determine the source code location, but as tests are executed relative to the directory where `_test.go` is located, so in `$SOURCEDIR/e2e`, `filepath.Dir(cwd)` is equivalent to source directory

I think I understand what you are saying.

I would not mind adding SOURCEDIR to buildcfg (I actually would need this to test the plugin stuff; the closest right now is buildcfg.BUILDDIR, which by default contains a directory one deeper than SOURCEDIR, but that can be changed). You could think of it as a debugging tool, it's pretty common to record that kind of information.

If the reason for this logic is getting to SOURCEDIR from e2e, I would prefer to make that change (add SOURCEDIR to buildcfg).